lib/DxcSupport/HLSLOptions.cpp

```diff
@@ -588,7 +588,35 @@ int ReadDxcOpts(const OptTable *optionTable, unsigned flagsToInclude,
   opts.UseInstructionNumbers = Args.hasFlag(OPT_Ni, OPT_INVALID, false);
   opts.UseInstructionByteOffsets = Args.hasFlag(OPT_No, OPT_INVALID, false);
   opts.UseHexLiterals = Args.hasFlag(OPT_Lx, OPT_INVALID, false);
-  opts.Preprocess = Args.getLastArgValue(OPT_P);
+  if (Args.hasFlag(OPT_P, OPT_INVALID, false)) {
+    // Default preprocess filename is InputName.i.
+    llvm::SmallString<128> Path(Args.getLastArgValue(OPT_INPUT));
+    llvm::sys::path::replace_extension(Path, "i");
+    // Try to get preprocess filename from Fi.
+    opts.Preprocess = Args.getLastArgValue(OPT_Fi, Path).str();
+    // Hack to support fxc style /P preprocess_filename.
+    // When there're more than 1 Input file, use the input which is after /P as
+    // preprocess.
+    std::vector<std::string> Inputs = Args.getAllArgValues(OPT_INPUT);
```

Comment:
I feel that the code below should only be run if there is no -Fi parameter. Where -P and -Fi is used such as `dxc -Fi out.pp -P in.hlsl`, I think that `in.hlsl` will get treated as the preprocessor output, which probably wasn't the intent.